src/plugins/data/common/index_patterns/fields/field_list.ts

```diff
@@ -18,95 +18,110 @@
   add(field: FieldSpec): void;
+  getByName(name: IndexPatternField['name']): IndexPatternField | undefined;
+  getByType(type: IndexPatternField['type']): IndexPatternField[];
   remove(field: IFieldType): void;
+  removeAll(): void;
+  replaceAll(specs: FieldSpec[]): void;
   update(field: FieldSpec): void;
 }
 
 export type CreateIndexPatternFieldList = (
-  indexPattern: IIndexPattern,
+  indexPattern: IndexPattern,
   specs?: FieldSpec[],
-  shortDotsEnable?: boolean
+  shortDotsEnable?: boolean,
+  onNotification?: OnNotification
 ) => IIndexPatternFieldList;
 
-export const getIndexPatternFieldListCreator = ({
-  fieldFormats,
-  onNotification,
-}: FieldListDependencies): CreateIndexPatternFieldList => (...fieldListParams) => {
-  class FieldList extends Array<Field> implements IIndexPatternFieldList {
-    private byName: FieldMap = new Map();
-    private groups: Map<Field['type'], FieldMap> = new Map();
-    private indexPattern: IIndexPattern;
-    private shortDotsEnable: boolean;
-    private setByName = (field: Field) => this.byName.set(field.name, field);
-    private setByGroup = (field: Field) => {
-      if (typeof this.groups.get(field.type) === 'undefined') {
-        this.groups.set(field.type, new Map());
-      }
-      this.groups.get(field.type)!.set(field.name, field);
-    };
-    private removeByGroup = (field: IFieldType) => this.groups.get(field.type)!.delete(field.name);
+export class FieldList extends Array<IndexPatternField> implements IIndexPatternFieldList {
+  private byName: FieldMap = new Map();
+  private groups: Map<IndexPatternField['type'], FieldMap> = new Map();
+  private indexPattern: IndexPattern;
+  private shortDotsEnable: boolean;
+  private onNotification: OnNotification;
+  private setByName = (field: IndexPatternField) => this.byName.set(field.name, field);
+  private setByGroup = (field: IndexPatternField) => {
+    if (typeof this.groups.get(field.type) === 'undefined') {
+      this.groups.set(field.type, new Map());
+    }
+    this.groups.get(field.type)!.set(field.name, field);
+  };
+  private removeByGroup = (field: IFieldType) => this.groups.get(field.type)!.delete(field.name);
+  private calcDisplayName = (name: string) =>
+    this.shortDotsEnable ? shortenDottedString(name) : name;
+  constructor(
+    indexPattern: IndexPattern,
+    specs: FieldSpec[] = [],
+    shortDotsEnable = false,
+    onNotification = () => {}
+  ) {
+    super();
+    this.indexPattern = indexPattern;
+    this.shortDotsEnable = shortDotsEnable;
+    this.onNotification = onNotification;
 
-    constructor(indexPattern: IIndexPattern, specs: FieldSpec[] = [], shortDotsEnable = false) {
-      super();
-      this.indexPattern = indexPattern;
-      this.shortDotsEnable = shortDotsEnable;
+    specs.map((field) => this.add(field));
+  }
 
-      specs.map((field) => this.add(field));
-    }
+  getByName = (name: IndexPatternField['name']) => this.byName.get(name);
```

Comment:
All these could have `readonly` modifier.

```suggestion
  readonly getByName = (name: IndexPatternField['name']) => this.byName.get(name);
```

or

```suggestion
  public readonly getByName = (name: IndexPatternField['name']) => this.byName.get(name);
```